File: java/src/days/Day1.java
package src.days;

import java.io.BufferedReader;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.stream.IntStream;

public class Day1 extends Day {
    @Override
    public String process(String inputFileName, boolean isPart2) {        
        Path filePath = Path.of("resources", inputFileName);
         try (BufferedReader reader = Files.newBufferedReader(filePath, StandardCharsets.UTF_8)) {        
            return isPart2 ? partTwo(reader) : partOne(reader);
        } catch (IOException e) {
            e.printStackTrace();
        } catch (Exception e) {
            e.printStackTrace();
        }
        return "Error";
    }

    private static String partOne(BufferedReader reader) throws IOException {
        String line;
        ArrayList<Integer> leftList = new ArrayList<>();
        ArrayList<Integer> rightList = new ArrayList<>();


        while ((line = reader.readLine()) != null) {
            String[] input = line.split("\\s+");
            leftList.add(Integer.parseInt(input[0]));
            rightList.add(Integer.parseInt(input[1]));                                
        }

        Collections.sort(leftList);
        Collections.sort(rightList);

        int similarity = IntStream.range(0, leftList.size())
            .map(i -> Math.abs(leftList.get(i) - rightList.get(i)))
            .sum();

        return String.valueOf(similarity);
    }

    private static String partTwo(BufferedReader reader) throws IOException {
        String line;

        ArrayList<Integer> leftList = new ArrayList<>();
        HashMap<Integer, Integer> rightMap = new HashMap<>();

        while ((line = reader.readLine()) != null) {
            String[] input = line.split("\\s+");            
            int rightValue = Integer.parseInt(input[1]);
            leftList.add(Integer.parseInt(input[0]));
            rightMap.putIfAbsent(rightValue,0);
            rightMap.put(rightValue,rightMap.get(rightValue) + 1);                                
        }

        int similarity = leftList.stream()
            .mapToInt(value -> value * rightMap.getOrDefault(value, 0))
            .sum();

        return String.valueOf(similarity);
    }
}

File: python/src/days/day1.py
import io
from pathlib import Path
from days.day import Day

class Day1(Day):
    
    def process(self, input_filename: str, is_part2: bool = False) -> str:
        try:
            file_path = Path("resources") / input_filename
            with open(file_path, "r", encoding="utf-8") as reader:
                return self.part_two(reader) if is_part2 else self.part_one(reader)

        except FileNotFoundError as e:
            print(f"Error: Input file '{input_filename}' not found.")
            return "Error"
        except Exception as e:
            print(f"An error occurred: {e}")
            return "Error"


    def part_one(self, reader) -> str:
        left_list = []
        right_list = []
        for line in reader:
            input_values = [int(value) for value in line.split()]
            left_list.append(input_values[0])
            right_list.append(input_values[1])

        left_list.sort()
        right_list.sort()

        similarity = sum(abs(left - right) for left, right in zip(left_list, right_list))
        return str(similarity)

    def part_two(self, reader) -> str:
        left_list= []
        right_map = {}
        for line in reader:
            input_values = [int(value) for value in line.split()]
            left_list.append(input_values[0])
            value = input_values[1]
            right_map[value] = right_map.get(value, 0) + 1

        similarity = sum(value * right_map.get(value, 0) for value in left_list)
        return str(similarity)
